### ISSUE TITLE
Torch Distributed Elastic has no way to address non-retriable errors

### ISSUE THREAD

**Issue by bgedik** [pytorch/pytorch#133877]
### 🚀 The feature, motivation and pitch

Looking at https://pytorch.org/docs/stable/elastic/errors.html, I don't see any way to avoid restarts when a non-retriable user error has happened. It is often important for user errors to have varieties as well. For instance, user's training module may discover an error at runtime that is non-retriable (say a node level issue that requires an orchestration level restart). There is currently no way to tell elastic agent that the exception should not be retried.  

### Alternatives

Wait until we run out of retries.

### Additional context

None

**Comment by tianyu-l** [comment]
Hey @d4l3k can you help answer?

**Comment by d4l3k** [comment]
I don't believe we have any way to do this currently. cc @kiukchung 

It's technically feasible to do. Would require:

1. adding a special error type to indicate a non-retryable error and propagate it through the error file https://github.com/pytorch/pytorch/blob/b567ca0f515f5910535cd5b9c25130a68b1e0836/torch/distributed/elastic/multiprocessing/errors/error_handler.py#L76
2. indicate to scheduler that things need to restart. The easiest way is to have the scheduler assume any failure is hard. If we need to do a distributed exit we'd need to close the rendezvous which may require a bit more work

You could also use a dirty hack and have the worker process SIGTERM the agent and thus trigger an orchestration level restart

**Comment by bgedik** [comment]
We worked around this by using a modified elastic agent. However, I am a bit surprised that this is not supported by PyTorch as the concept of non-retriable errors are very fundamental.

**Comment by fduwjj** [comment]
@bgedik you are more than welcomed to improve pytorch to sending out a PR to make it better.

**Comment by bgedik** [comment]
This requires a mini design in my opinion. If there is someone who is more familiar with the internals of PyTorch distributed, who is willing to pair with me, I don't mind putting cycles on it.
